test/tls13secretstest.c

```diff
@@ -0,0 +1,348 @@
+0x74, 0x86, 0x88, 0xe9, 0x7f, 0x72, 0xfb, 0xf3, 0x33, 0x1e, 0xfb, 0x55
+};
+
+static const char *server_ats_label = "server application traffic secret";
+
+static unsigned char server_ats[] = {
+0xd1, 0xbf, 0xdc, 0x8b, 0x84, 0xf4, 0x16, 0xb7, 0xc6, 0x90, 0xd9, 0xc9, 0x2c,
+0x23, 0x11, 0xb3, 0x05, 0xad, 0x75, 0xfc, 0xe6, 0x29, 0x90, 0x2b, 0xe1, 0x03,
+0xdd, 0x0c, 0x12, 0x51, 0xea, 0xd2,
+};
+
+static unsigned char server_ats_key[] = {
+0x35, 0xc2, 0xd1, 0x54, 0xa8, 0x43, 0x03, 0xc6, 0x55, 0xa0, 0x2e, 0x5e, 0x1f,
+0x82, 0x31, 0x62
+};
+
+static unsigned char server_ats_iv[] = {
+0xe5, 0x77, 0xd9, 0x8a, 0xb3, 0x2e, 0xec, 0x79, 0xb1, 0x63, 0x68, 0xc2
+};
+
+/* Mocked out implementations of various functions */
+int ssl3_digest_cached_records(SSL *s, int keep)
+{
+    return 1;
+}
+
+static int full_hash = 0;
+
+/* Give a hash of the currently set handshake */
+int ssl_handshake_hash(SSL *s, unsigned char *out, size_t outlen,
+                       size_t *hashlen)
+{
+    if (sizeof(hs_start_hash) > outlen
+            || sizeof(hs_full_hash) != sizeof(hs_start_hash))
+        return 0;
+
+    if (full_hash) {
+        memcpy(out, hs_full_hash, sizeof(hs_full_hash));
+        *hashlen = sizeof(hs_full_hash);
+    } else {
+        memcpy(out, hs_start_hash, sizeof(hs_start_hash));
+        *hashlen = sizeof(hs_start_hash);
+    }
+
+    return 1;
+}
+
+const EVP_MD *ssl_handshake_md(SSL *s)
+{
+    return EVP_sha256();
+}
+
+/* End of mocked out code */
+
+static int test_secret(SSL *s, unsigned char *prk,
+                       const unsigned char *label, size_t labellen,
+                       const unsigned char *ref_secret,
+                       const unsigned char *ref_key, const unsigned char *ref_iv)
+{
+    size_t hashsize = EVP_MD_size(ssl_handshake_md(s));
+    unsigned char gensecret[EVP_MAX_MD_SIZE];
+    unsigned char key[KEYLEN];
+    unsigned char iv[IVLEN];
+
+    if (!tls13_derive_secret(s, prk, label, labellen, gensecret)) {
+        printf("Secret generation failed\n");
+        return 0;
+    }
+
+    if (memcmp(gensecret, ref_secret, hashsize) != 0) {
+        printf("Generated secret does not match\n");
+        return 0;
+    }
+
+    if (!tls13_derive_key(s, gensecret, key, KEYLEN)) {
+        printf("Key generation failed\n");
+        return 0;
+    }
+
+    if (memcmp(key, ref_key, KEYLEN) != 0) {
+        printf("Generated key does not match\n");
+        return 0;
+    }
+
+    if (!tls13_derive_iv(s, gensecret, iv, IVLEN)) {
+        printf("IV generation failed\n");
+        return 0;
+    }
+
+    if (memcmp(iv, ref_iv, IVLEN) != 0) {
+        printf("Generated IV does not match\n");
+        return 0;
+    }
+
+    return 1;
+}
+
+static int test_handshake_secrets(void)
+{
+    SSL_CTX *ctx;
+    SSL *s;
+    int ret = 0;
+    size_t hashsize;
+    unsigned char out_master_secret[EVP_MAX_MD_SIZE];
+    size_t master_secret_length;
+
+    ctx = SSL_CTX_new(TLS_method());
+
```

Comment:
Blank line deleted, NULL initialisers added and goto err. Not merged the assignment into the if...I really hate that style and try to avoid it wherever possible (and remove it when I can too).